src/test/java/org/apache/zookeeper/book/TestTaskAssignment.java

```diff
@@ -286,4 +289,57 @@ public void taskZooKeeperCrash() throws Exception {
         w3.close();
         c.close();
     }
+
+    @Test(timeout=50000)
+    public void taskWorkerAssignmentCallback() throws Exception {
+        LOG.info("Starting master - Sequential");
+        Master m = new Master("localhost:" + port);
+        m.startZK();
+
+        while(!m.isConnected()){
+            Thread.sleep(500);
+        }
+
+        m.bootstrap();
+        m.runForMaster();
+
+        while(m.getState() == MasterStates.RUNNING){
+            Thread.sleep(100);
+        }
+
+        LOG.info("Starting worker");
+        Worker w = new Worker("localhost:" + port);
+        w.startZK();
+        while(!w.isConnected()){
+            Thread.sleep(100);
+        }
+
+        /*
+         * bootstrap() create some necessary znodes.
+         */
+        w.bootstrap();
+        /*
+         * Registers this worker so that the leader knows that
+         * it is here.
+         */
+        w.register();
+
+        LOG.info("Starting client");
+        Client c = new Client("localhost:" + port);
+        c.startZK();
+
+        while(!c.isConnected() &&
```

Comment:
Not that this is relevant to this test but this condition seems wrong. The loop exits if either the **client connects**, or the **worker disconnects**, which makes the condition logically meaningless and couples two unrelated connection states. If the intention was to wait for both client and worker to connect before proceeding with the test, following logic makes more sense:
```java
while (!c.isConnected()) {
    Thread.sleep(100);
}

while (!w.isConnected()) {
    Thread.sleep(100);
}
```